File: backend/tests/experiments/attribution.test.ts
/**
 * Tests for `experiments/attribution.ts` — P&L decomposition into
 * priceImpact, fees, gasCost, and rebates.
 *
 * Pure math, no mocks.  Cases cover:
 *  - simple positive attribution (sum equals total)
 *  - zero IL / zero fees → net equals sum of remaining components
 *  - fees + rebates + price impact all positive
 *  - negative P&L (loss) — total below zero
 *  - all-zero components
 *  - components echoed in the breakdown
 */
import { describe, it, expect } from 'vitest';

import {
  runAttributionExperiment,
  attributeProfit,
  type AttributionComponents,
} from '../../src/experiments/attribution.js';

const ONE_E18 = 10n ** 18n;

describe('experiments/attribution', () => {
  describe('attributeProfit (helper)', () => {
    it('total = priceImpact + fees - gasCost + rebates', () => {
      const r = attributeProfit({
        priceImpact: 100n * ONE_E18,
        fees: 5n * ONE_E18,
        gasCost: 2n * ONE_E18,
        rebates: 1n * ONE_E18,
      });
      expect(r.total).toBe(104n * ONE_E18);
    });

    it('all-zero components ⇒ total is 0 and percentages are 0', () => {
      const r = attributeProfit({ priceImpact: 0n, fees: 0n, gasCost: 0n, rebates: 0n });
      expect(r.total).toBe(0n);
      expect(r.percentages).toEqual({ priceImpact: 0, fees: 0, gasCost: 0, rebates: 0 });
    });

    it('handles negative P&L (loss)', () => {
      const r = attributeProfit({
        priceImpact: -10n * ONE_E18,
        fees: 0n,
        gasCost: 5n * ONE_E18,
        rebates: 0n,
      });
      expect(r.total).toBe(-15n * ONE_E18);
    });

    it('breakdown echoes input values', () => {
      const c: AttributionComponents = {
        priceImpact: 7n * ONE_E18,
        fees: 1n * ONE_E18,
        gasCost: 2n * ONE_E18,
        rebates: 0n,
      };
      const r = attributeProfit(c);
      expect(r.breakdown).toEqual(c);
    });

    it('uses absolute total for percentage denominator', () => {
      const r = attributeProfit({
        priceImpact: -20n * ONE_E18,
        fees: 0n,
        gasCost: 5n * ONE_E18,
        rebates: 0n,
      });
      expect(r.percentages.gasCost).toBeCloseTo(0.2, 6);
      expect(Math.abs(r.percentages.priceImpact)).toBeCloseTo(0.8, 6);
    });
  });

  describe('runAttributionExperiment', () => {
    it('net = priceImpact + fees - gasCost + rebates', () => {
      const res = runAttributionExperiment({
        reserve0: 80_000n * ONE_E18,
        reserve1: 160_000_000n * 10n ** 6n,
        amountIn: 10n * ONE_E18,
        fee: 3000,
        rebates: 0n,
        gasCost: 0n,
      });
      const r = res.result;
      // Result uses string bigints via the .toString() convention.
      const net = BigInt(r.netPnl as string);
      const priceImpact = BigInt(r.priceImpact as string);
      const fees = BigInt(r.fees as string);
      const gasCost = BigInt(r.gasCost as string);
      const rebates = BigInt(r.rebates as string);
      expect(net).toBe(priceImpact + fees - gasCost + rebates);
    });

    it('all positive components → net > 0 and equals sum of components', () => {
      const res = runAttributionExperiment({
        reserve0: 80_000n * ONE_E18,
        reserve1: 160_000_000n * 10n ** 6n,
        amountIn: 10n * ONE_E18,
        fee: 3000,
        rebates: 1n * 10n ** 6n, // 1 USDC
        gasCost: 5n * 10n ** 14n, // 0.0005 ETH
      });
      const r = res.result;
      const net = BigInt(r.netPnl as string);
      const fees = BigInt(r.fees as string);
      const rebates = BigInt(r.rebates as string);
      const gasCost = BigInt(r.gasCost as string);
      // net > 0 because price impact (negative) is small relative to rebates.
      // More importantly, the equation holds:
      const priceImpact = BigInt(r.priceImpact as string);
      expect(net).toBe(priceImpact + fees - gasCost + rebates);
      // And fees are negative (a cost from the trader's perspective).
      expect(fees < 0n).toBe(true);
      // Rebates are positive.
      expect(rebates > 0n).toBe(true);
    });

    it('zero fee case → fees component is 0', () => {
      const res = runAttributionExperiment({
        reserve0: 80_000n * ONE_E18,
        reserve1: 160_000_000n * 10n ** 6n,
        amountIn: 10n * ONE_E18,
        fee: 0,
      });
      const r = res.result;
      expect(BigInt(r.fees as string)).toBe(0n);
      // Net still equals the other components.
      const net = BigInt(r.netPnl as string);
      const priceImpact = BigInt(r.priceImpact as string);
      const gasCost = BigInt(r.gasCost as string);
      const rebates = BigInt(r.rebates as string);
      expect(net).toBe(priceImpact - gasCost + rebates);
    });

    it('zero fee case → net = priceImpact - gasCost + rebates', () => {
      // Use 0% pool so the fees component is forced to 0; gas/rebates
      // are also 0 by default, so net should equal priceImpact exactly.
      const res = runAttributionExperiment({
        reserve0: 80_000n * ONE_E18,
        reserve1: 160_000_000n * 10n ** 6n,
        amountIn: 10n * ONE_E18,
        fee: 0,
      });
      const r = res.result;
      expect(BigInt(r.fees as string)).toBe(0n);
      const net = BigInt(r.netPnl as string);
      const priceImpact = BigInt(r.priceImpact as string);
      const gasCost = BigInt(r.gasCost as string);
      const rebates = BigInt(r.rebates as string);
      expect(net).toBe(priceImpact - gasCost + rebates);
    });

    it('zero fee and zero gas/rebates → net = priceImpact', () => {
      const res = runAttributionExperiment({
        reserve0: 80_000n * ONE_E18,
        reserve1: 160_000_000n * 10n ** 6n,
        amountIn: 10n * ONE_E18,
        fee: 0,
      });
      const r = res.result;
      const net = BigInt(r.netPnl as string);
      const priceImpact = BigInt(r.priceImpact as string);
      expect(net).toBe(priceImpact);
    });

    it('returns a non-negative durationMs', () => {
      const res = runAttributionExperiment({
        reserve0: ONE_E18,
        reserve1: ONE_E18,
        amountIn: ONE_E18,
        fee: 3000,
      });
      expect(res.durationMs).toBeGreaterThanOrEqual(0);
    });

    it('echoes the input reserves in the result', () => {
      const r0 = 80_000n * ONE_E18;
      const r1 = 160_000_000n * 10n ** 6n;
      const res = runAttributionExperiment({
        reserve0: r0,
        reserve1: r1,
        amountIn: 10n * ONE_E18,
        fee: 3000,
      });
      expect(res.result.reserve0).toBe(r0.toString());
      expect(res.result.reserve1).toBe(r1.toString());
      expect(res.result.feeHundredthsBip).toBe(3000);
    });

    it('result contains all four attribution keys', () => {
      const res = runAttributionExperiment({
        reserve0: ONE_E18,
        reserve1: ONE_E18,
        amountIn: ONE_E18,
        fee: 3000,
      });
      const r = res.result;
      expect(r).toHaveProperty('priceImpact');
      expect(r).toHaveProperty('fees');
      expect(r).toHaveProperty('gasCost');
      expect(r).toHaveProperty('rebates');
      expect(r).toHaveProperty('netPnl');
    });
  });
});

File: backend/tests/experiments/attribution_test.py
"""Unit tests for `dtm_backend.experiments.attribution`.

Attribution decomposes a single LP position's return over a
period into 4 components:

  1. **Fees** earned from swap fees (always ≥ 0)
  2. **IL** (impermanent loss; always ≤ 0)
  3. **Rebates** / incentives (always ≥ 0)
  4. **Net PnL** = fees + IL + rebates

The wire format is the one consumed by
`POST /api/v1/experiments/attribution` and by
`HttpAPI.runAttributionExperiment` on the frontend::

    {
      "durationMs": <int>,
      "result": {
        "netPnl": "<decimal string of raw integer>",
      }
    }

The "raw integer" is the net PnL in the same unit as the
input reserves (e.g. wei for ETH, 1e6 for USDC).  The
*components* themselves are not exposed on the wire —
they are an internal accounting device that sums to netPnl.
"""
from __future__ import annotations

from dtm_backend.experiments.attribution import (
    AttributionInput,
    run_attribution,
)


def test_attribution_net_pnl_is_sum_of_components() -> None:
    """Net PnL = fees + IL + rebates (algebraic sum)."""
    res = run_attribution(
        AttributionInput(
            reserve0=80_000 * 10**18,
            reserve1=160_000_000 * 10**6,
            amount_in=1_000 * 10**18,
            fee_bip=3000,
            # Fictional but bounded scenario.
            price_ratio=1.5,
            fees=10**18,        # 1 ETH of fees
            rebates=5 * 10**17, # 0.5 ETH of incentives
        )
    )
    # The implementation must round-trip its own internal sum.
    assert res.net_pnl == res.fees + res.il + res.rebates


def test_attribution_negative_il_reduces_net_pnl() -> None:
    """A more severe IL ⇒ strictly lower net PnL (with fees/rebates held constant)."""
    base = dict(
        reserve0=80_000 * 10**18,
        reserve1=160_000_000 * 10**6,
        amount_in=1_000 * 10**18,
        fee_bip=3000,
        fees=10**18,
        rebates=0,
    )
    a = run_attribution(AttributionInput(**base, price_ratio=1.1)).net_pnl
    b = run_attribution(AttributionInput(**base, price_ratio=1.5)).net_pnl
    c = run_attribution(AttributionInput(**base, price_ratio=2.0)).net_pnl
    # A bigger price move ⇒ more negative IL ⇒ smaller net PnL.
    assert a > b > c


def test_attribution_higher_fees_increase_net_pnl() -> None:
    """More fees ⇒ strictly higher net PnL (other things equal)."""
    base = dict(
        reserve0=80_000 * 10**18,
        reserve1=160_000_000 * 10**6,
        amount_in=1_000 * 10**18,
        fee_bip=3000,
        price_ratio=1.5,
        rebates=0,
    )
    a = run_attribution(AttributionInput(**base, fees=10**18)).net_pnl
    b = run_attribution(AttributionInput(**base, fees=2 * 10**18)).net_pnl
    c = run_attribution(AttributionInput(**base, fees=10**19)).net_pnl
    assert a < b < c


def test_attribution_rebates_increase_net_pnl() -> None:
    """Rebates are a positive contribution (subject to the IL cost)."""
    base = dict(
        reserve0=80_000 * 10**18,
        reserve1=160_000_000 * 10**6,
        amount_in=1_000 * 10**18,
        fee_bip=3000,
        price_ratio=1.5,
        fees=10**18,
    )
    a = run_attribution(AttributionInput(**base, rebates=0)).net_pnl
    b = run_attribution(AttributionInput(**base, rebates=10**18)).net_pnl
    assert b > a
    assert b - a == 10**18


def test_attribution_zero_fees_and_rebates_yields_negative_net_pnl() -> None:
    """Without any fees or rebates, net PnL must be negative (IL-only)."""
    res = run_attribution(
        AttributionInput(
            reserve0=80_000 * 10**18,
            reserve1=160_000_000 * 10**6,
            amount_in=1_000 * 10**18,
            fee_bip=3000,
            price_ratio=1.5,
            fees=0,
            rebates=0,
        )
    )
    assert res.net_pnl < 0
    # Net PnL is exactly the (negative) IL component.
    assert res.net_pnl == res.il
    assert res.fees == 0
    assert res.rebates == 0


def test_attribution_rejects_non_positive_price_ratio() -> None:
    """``price_ratio`` must be > 0 (we take a log of it)."""
    import pytest

    with pytest.raises(ValueError, match="price_ratio"):
        run_attribution(
            AttributionInput(
                reserve0=10**18,
                reserve1=10**6,
                amount_in=10**18,
                fee_bip=3000,
                price_ratio=0,
                fees=0,
                rebates=0,
            )
        )
